src/v/cloud_storage/topic_manifest.cc

```diff
@@ -21,12 +23,175 @@
+            } else if ("compression" == _key) {
+                _topic_config.properties.compression
+                  = boost::lexical_cast<model::compression>(
+                    std::string(str, length));
+            } else if ("cleanup_policy_bitflags" == _key) {
+                _topic_config.properties.cleanup_policy_bitflags
+                  = boost::lexical_cast<model::cleanup_policy_bitflags>(
+                    std::string(str, length));
+            } else if ("compaction_strategy" == _key) {
+                _topic_config.properties.compaction_strategy
+                  = boost::lexical_cast<model::compaction_strategy>(
+                    std::string(str, length));
+            } else if ("timestamp_type" == _key) {
+                _topic_config.properties.timestamp_type
+                  = boost::lexical_cast<model::timestamp_type>(
+                    std::string(str, length));
+            } else {
+                return false;
+            }
+            if (_required_fields.contains(_key)) {
+                _seen_required_fields.insert(_key);
+            }
+            _state = all_required_complete() ? expectKeyOrObjectEnd : expectKey;
+            return true;
+        default:
+            return false;
+        }
+    }
+
+    bool Uint(int u) {
+        vlog(cst_log.trace, "Uint({}), state: {}", u, _state);
+        switch (_state) {
+        case expectValue:
+            if ("version" == _key) {
+                if (u != static_cast<int>(topic_manifest_version::v1)) {
+                    throw std::runtime_error(
+                      "topic manifest version not supported");
+                }
+            } else if ("partition_count" == _key) {
+                _topic_config.partition_count = u;
+            } else if ("replication_factor" == _key) {
+                _topic_config.replication_factor = u;
+            } else if ("revision_id" == _key) {
+                _revision_id = u;
+            } else if ("segment_size" == _key) {
+                _topic_config.properties.segment_size = u;
+            } else if ("retention_bytes" == _key) {
+                _topic_config.properties.retention_bytes = tristate<size_t>(u);
+            } else if ("retention_duration" == _key) {
+                _topic_config.properties.retention_duration
+                  = tristate<std::chrono::milliseconds>(
+                    std::chrono::milliseconds(u));
+            } else {
+                return false;
+            }
+            if (_required_fields.contains(_key)) {
+                _seen_required_fields.insert(_key);
+            }
+            _state = all_required_complete() ? expectKeyOrObjectEnd : expectKey;
+            return true;
+        default:
+            return false;
+        }
+    }
+
+    bool EndObject(rapidjson::SizeType size) {
+        vlog(cst_log.trace, "EndObject({}), state_: {}", size, _state);
+        return _state == expectKeyOrObjectEnd;
```

Comment:
The `expectKeyOrObjectEnd` state feels a little awkward, I wonder if it would be simpler to grok the state machine if this state was removed and then:
```suggestion
        return all_required_complete();
```
The parser already checks that keys and values are paired.